include/ppx/application.h

```diff
@@ -307,6 +311,11 @@ struct ApplicationSettings
             uint32_t graphicsQueueCount = 1;
             uint32_t computeQueueCount  = 0;
             uint32_t transferQueueCount = 0;
+
+            // Enable support for this shading rate mode on the device.
+            // The application must not use FDM or VRS without setting this to
+            // the corresponding shading rate mode.
+            grfx::ShadingRateMode supportShadingRateMode = grfx::SHADING_RATE_NONE;
```

Comment:
This sets the default which the knobs can override. If the user doesn't specify any of the knob flags, the application-specified default will be used.

I think both an app-specific default and knobs to override that default are important.

For most apps, `NONE` is correct, but not for apps specifically designed to use FDM or VRS.

Being overridable via knobs is also important to be able to switch the shading rate mode without rebuilding the application (and this is what I've been using for the foveation benchmarking).

Reply:
Ok, fair enough. There is some weirdness right now in that you can specify a default value for the knob for an application which would always override the settings value, but since you are not using knob default it probably doesn't matter. 
Would a specific app error out if the shading rate mode is not supported? Say if the app can only support FDM and VRS is requested.

Reply:
> Would a specific app error out if the shading rate mode is not supported? Say if the app can only support FDM and VRS is requested.

This flag just enables the shading rate mode on the device, but doesn't attempt to change the render pass to use the shading rate mode.

I think the use of the setting and knob might be more clear with an example from the foveation benchmark app in my [experimental-foveation branch](https://github.com/bjoeris/bigwheels/blob/experimental-foveation/benchmarks/foveation/main.cpp):

- In `FoveationBenchmarkApp::Config()` ([benchmarks/foveation/main.cpp:256](https://github.com/bjoeris/bigwheels/blob/ea601395784ddf21860f2f9d7902fb985699e424/benchmarks/foveation/main.cpp#L256)), the default shading rate mode is set in `ApplicationSettings`.
  - Most applications skip this step, so the default shading rate mode would be `SHADING_RATE_NONE`.
- In `Application::UpdateStandardSettings()` ([src/ppx/application.cpp:1147](https://github.com/bjoeris/bigwheels/blob/ea601395784ddf21860f2f9d7902fb985699e424/src/ppx/application.cpp#L1147)), the default shading rate mode is overwritten if the `--shading-rate-mode` flag is used.
- In `Application::InitializeGrfxDevice()` ([src/ppx/application.cpp:354](https://github.com/bjoeris/bigwheels/blob/ea601395784ddf21860f2f9d7902fb985699e424/src/ppx/application.cpp#L354)), the default or overwritten shading rate mode is passed to the `DeviceCreateInfo`. This will cause an error if the device does not support this shading rate mode, or is using the DX12 backend.
- In `vk::Device::ConfigureShadingRateCapabilities()` ([src/ppx/grfx/vk/vk_device.cpp:320](https://github.com/bjoeris/bigwheels/blob/ea601395784ddf21860f2f9d7902fb985699e424/src/ppx/grfx/vk/vk_device.cpp#L320)), the shading rate mode passed to the `DeviceCreateInfo` is saved in the `ShadingRateCapabilities` for the device.
- In `FoveationBenchmarkApp::Setup()` ([benchmarks/foveation/main.cpp:524](https://github.com/bjoeris/bigwheels/blob/ea601395784ddf21860f2f9d7902fb985699e424/benchmarks/foveation/main.cpp#L524)), the application uses the shading rate mode from the device's `ShadingRateCapabilities` to configure the render pass and pipeline.
  - Most applications skip this step, so their render pass wont actually use VRS or FDM, even if it is enabled on the device.

Reply:
sg, thanks for the detailed walkthrough!